### ISSUE TITLE
tfsec hangs indefinitely on MacOS 10.14.6 (Mojave)

### ISSUE THREAD

**Issue by academ1c** [liamg/tfsec#52]
I runned test on MacOS from scratch and from release file (github.com)
Test runned more 40 mins and not ended (
Check please
I am using modules also in my TF files (tf 0.12.10)

**Comment by liamg** [comment]
Thanks for the report, that sounds very odd! What are the full options you're providing tfsec, if any? And how big is the source code you're scanning?

**Comment by academ1c** [comment]
1. I build from scratch  = /path/to/bin/tfsec  /path/to/tf/project
2. I downloaded to /usr/local/bin//tsfec = chmod +x  -> tsfec /path/to/tf/project

Result - same

**Comment by liamg** [comment]
Hmm, I'll add a `--verbose` option soon so we can examine what's going on in that case.

**Comment by giany** [comment]
I don't think it's related to modules. I have none and when I run it it still hangs. Noticed also that CPU usage increases.

**Comment by liamg** [comment]
Thanks for the extra info @giany - do you think you could paste the output of `tree` in the directory you're running the tool in? I can't seem to reproduce this with the ideas I've had so far.

**Comment by giany** [comment]
```
mbp:terraform giany$ ls -al
total 96
drwxr-xr-x  15 giany  staff   480 Nov 13 15:17 .
drwxr-xr-x   7 giany  staff   224 Oct  1 22:56 ..
drwxr-xr-x   4 giany  staff   128 Nov 13 15:17 .terraform
-rw-r--r--   1 giany  staff   501 Oct  2 13:24 acm.tf
-rw-r--r--   1 giany  staff   796 Oct  2 13:19 alb.tf
-rw-r--r--   1 giany  staff  1197 Nov  6 14:36 asg.tf
-rw-r--r--   1 giany  staff  1151 Oct  2 13:48 iam.tf
-rw-r--r--   1 giany  staff   702 Oct  2 13:53 locals.tf
-rw-r--r--   1 giany  staff   306 Oct  2 13:19 main.tf
-rw-r--r--   1 giany  staff   111 Oct  2 13:19 output.tf
-rw-r--r--   1 giany  staff   107 Sep 24 23:09 provider.tf
-rw-r--r--   1 giany  staff   157 Sep 24 22:41 remote.tf
-rw-r--r--   1 giany  staff  1560 Oct  2 13:19 sg.tf
-rw-r--r--   1 giany  staff   223 Oct  2 13:19 variables.tf
-rw-r--r--   1 giany  staff   457 Sep 24 23:32 vpc.tf
mbp:terraform giany$ tree
.
├── acm.tf
├── alb.tf
├── asg.tf
├── iam.tf
├── locals.tf
├── main.tf
├── output.tf
├── provider.tf
├── remote.tf
├── sg.tf
├── variables.tf
└── vpc.tf

0 directories, 12 files
```

Btw if I remove the .terraform dir it haves properly.

**Comment by liamg** [comment]
Hmm, adding a .terraform dir doesn't let me reproduce this unfortunately - it must be something to do with the contents of that directory.

**Comment by academ1c** [comment]
.terraform # tree | wc -l
4419
))))

**Comment by giany** [comment]
For this repo I have 5 modules. So in .terraform/modules :
```
ls 
acm		alb		asg		modules.json	sg		vpc

mbp:modules giany$ cat modules.json
{"Modules":[{"Key":"asg","Source":"terraform-aws-modules/autoscaling/aws","Version":"3.4.0","Dir":".terraform/modules/asg/terraform-aws-modules-terraform-aws-autoscaling-07426a1"},{"Key":"sg","Source":"terraform-aws-modules/security-group/aws","Version":"3.1.0","Dir":".terraform/modules/sg/terraform-aws-modules-terraform-aws-security-group-d55e4de"},{"Key":"vpc","Source":"terraform-aws-modules/vpc/aws","Version":"2.15.0","Dir":".terraform/modules/vpc/terraform-aws-modules-terraform-aws-vpc-13fc020"},{"Key":"","Source":"","Dir":"."},{"Key":"acm","Source":"terraform-aws-modules/acm/aws","Version":"2.4.0","Dir":".terraform/modules/acm/terraform-aws-modules-terraform-aws-acm-963ff73"},{"Key":"alb","Source":"terraform-aws-modules/alb/aws","Version":"4.1.0","Dir":".terraform/modules/alb/terraform-aws-modules-terraform-aws-alb-61e7316"}]}```

All are modules from terraform.

**Comment by liamg** [comment]
Perfect, thanks, I'll try and replicate this later on tonight.

**Comment by schosterbarak** [comment]
I've encountered the same

**Comment by liamg** [comment]
Easiest solution for now is to ignore dotfiles (#63). I want to support modules stored in `.terraform` soon, so I'll tackle the full problem then, hopefully this will fix it for now.

**Comment by schosterbarak** [comment]
Thanks for the fix @liamg . I've tried the same with latest release, it seems it still hangs.

**Comment by schosterbarak** [comment]
I think that what i'm seeing is infinite appending of `blocks` to a variable named `localBlocks` at https://github.com/liamg/tfsec/blob/9a5eb2e5fddd13cdde0b2fd9612b185eadcc264e/internal/app/tfsec/parser/parser.go#L158

**Comment by liamg** [comment]
@schosterbarak Awesome, nice find! Sorry I haven't had much time for this lately - I've just pushed a PR to fix that bug, if you have time to build from that branch and try again that'd be awesome.

**Comment by schosterbarak** [comment]
thanks @liamg !  i've just built from that branch and tried it. seems like the number of `childModules` at https://github.com/liamg/tfsec/blob/liamg-fix-bug-in-module-block-parsing/internal/app/tfsec/parser/parser.go#L209 is infinite  (len: 16882693) 
with a most of it as `nil` values. hope it helps.

**Comment by look4regev** [comment]
Thank you for reporting and working on this issue. It's really getting in the way of being able to use this awesome tool. 
I've made a simple script to workaround the problem in the meantime:
```
#!/bin/bash
# Running on each dir specifically
had_any_failure=0
for dir in $(find . -not -path '*/\.*' ! -path . -maxdepth 1 -type d); do
    echo "$dir"
    tfsec $dir
    if [ $? -gt 0 ]; then
      had_any_failure=1
    fi
done
exit $had_any_failure
```

**Comment by skyzyx** [comment]
TLDR; It would appear that `tfsec` **doesn't support modules**.

My debugging and sleuthing steps are below:

----

I'm going to add more information in hopes that the bug can be identified and fixed.

* **OS:** macOS 10.15.3 “Catalina”
* **tfsec:** v0.19.0

My Terraform code is a module, that will be imported and used by others. Here is `tree -a` from a fresh clone:

```
.
├── .circleci
│   └── config.yml
├── .editorconfig
├── .git
│   ├── HEAD
│   ├── config
│   ├── hooks
│   ├── index
│   ├── logs
│   │   ├── HEAD
│   │   └── refs
│   │       ├── heads
│   │       │   └── master
│   │       └── remotes
│   │           └── origin
│   │               └── HEAD
│   ├── objects
│   │   ├── info
│   │   └── pack
│   │       ├── pack-4c08ece7129b7f2a5b368aaf01f5f596abf307cd.idx
│   │       └── pack-4c08ece7129b7f2a5b368aaf01f5f596abf307cd.pack
│   ├── packed-refs
│   └── refs
│       ├── heads
│       │   └── master
│       ├── remotes
│       │   └── origin
│       │       └── HEAD
│       └── tags
├── .gitignore
├── .golangci.yml
├── .markdownlintrc
├── .templates
│   ├── Makefile
│   ├── Makefile.tmpl
│   └── README.md
├── .terraform-validator.yaml
├── .tflint.hcl
├── CHANGELOG.md
├── Makefile
├── README.md
├── VERSION
├── alb.tf
├── bin
│   ├── examples.sh
│   └── show-skips.sh
├── examples
│   └── aws_alb_test_example.tf
├── iam.tf
├── listener.tf
├── output.tf
├── package-lock.json
├── package.json
├── target_group.tf
├── tests
│   ├── .gitignore
│   ├── aws_alb_test.go
│   ├── aws_alb_test.tf
│   ├── go.mod
│   └── go.sum
├── tf-modules-stable.png
├── variables.tf
└── versions.tf
```

Nothing particularly fancy. At this point, running `tfsec .` in this directory will hang indefinitely.

----

Next, let's remove the _directories_ which start with a period.

```bash
rm -Rf .circleci/ .git/ .templates/
```

Running `tfsec .` in this directory will still hang indefinitely.

Next, let's remove the _files_ which start with a period.

```bash
rm -f .editorconfig .gitignore .golangci.yml .markdownlintrc .terraform-validator.yaml .tflint.hcl
```

Running `tfsec .` in this directory will still hang indefinitely.

----

Now, let's look at the current state with `tree -a`.

```
.
├── CHANGELOG.md
├── Makefile
├── README.md
├── VERSION
├── alb.tf
├── bin
│   ├── examples.sh
│   └── show-skips.sh
├── examples
│   └── aws_alb_test_example.tf
├── iam.tf
├── listener.tf
├── output.tf
├── package-lock.json
├── package.json
├── target_group.tf
├── tests
│   ├── .gitignore
│   ├── aws_alb_test.go
│   ├── aws_alb_test.tf
│   ├── go.mod
│   └── go.sum
├── tf-modules-stable.png
├── variables.tf
└── versions.tf

3 directories, 22 files
```

I notice one second-level file which begins with a period inside the `tests/` directory — `.gitignore`. Let's remove it.

```bash
rm -f tests/.gitignore
```

Running `tfsec .` in this directory will still hang indefinitely.

----

Not much left to whittle-down. Next I'm going to try running it from inside the `tests/` directory to see if we get anywhere.

```bash
$ cd tests/
$ tfsec .
```

Still hanging. I know that `aws_alb_test.tf` references _other modules_ which have not yet been downloaded. Let's show a redacted version of the file:

```hcl
# https://www.terraform.io/docs/configuration/terraform.html
terraform {
  required_version = ">= 0.12"
}

provider "aws" {
  # https://github.com/terraform-providers/terraform-provider-aws/releases/latest
  version = ">= 2.47, <3.0"
}

module "network" {
  source = "git::ssh://git@PRIVATE/terraform/network.git?ref=2.0.2"
}

module "aws_resource_tags" {
  source      = "git::ssh://git@PRIVATE/terraform/aws-resource-tags.git?ref=2.1.0"
  account     = "test"
  application = "sample-alb-test"
  environment = "dev"
  function    = "api"
  platform    = "test"
}

module "aws_base_ami" {
  source      = "git::ssh://git@PRIVATE/terraform/aws-base-ami.git?ref=1.0.2"
  ami_type    = "ecs"
  ami_version = "1.2.3"
}

module "alb_http" {
  source = "../"

  # Naming
  application_name = module.aws_resource_tags.application
  environment      = module.aws_resource_tags.environment

  # General configuration
  deletion_protection = false
  enable_ipv6         = false
  external            = false

  # Tags
  tags = module.aws_resource_tags.common_tags

  # Network/VPC configuration
  vpc_id  = module.network.vpc_id
  subnets = module.network.internal_subnet_ids
  security_groups = [
    module.network.default_security_group_id,
    module.network.web_vpc_security_group_id,
  ]
}
```

Let's see if `terraform init` helps with that.

```bash
$ terraform init

Initializing modules...
- alb_http in ..
Downloading git::ssh://git@PRIVATE/terraform/aws-base-ami.git?ref=1.0.2 for aws_base_ami...
- aws_base_ami in .terraform/modules/aws_base_ami
Downloading git::ssh://git@PRIVATE/terraform/aws-resource-tags.git?ref=2.1.0 for aws_resource_tags...
- aws_resource_tags in .terraform/modules/aws_resource_tags
Downloading git::ssh://git@PRIVATE/terraform/network.git?ref=2.0.2 for network...
- network in .terraform/modules/network

Initializing the backend...

Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "aws" (hashicorp/aws) 2.52.0...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

Still hanging. There is a…

```hcl
module "alb_http" {
  source = "../"
```

…which means the Terraform isn't entirely in this directory.

----

Let's try going back to the root directory, getting rid of the directories with _external module references_, and initializing Terraform.

```bash
cd ..
rm -Rf examples/ tests/
```

```bash
$ terraform init

Initializing the backend...

Initializing provider plugins...
- Checking for available provider plugins...
- Downloading plugin for provider "aws" (hashicorp/aws) 2.52.0...

The following providers do not have any version constraints in configuration,
so the latest version was installed.

To prevent automatic upgrades to new major versions that may contain breaking
changes, it is recommended to add version = "..." constraints to the
corresponding provider blocks in configuration, with the constraint strings
suggested below.

* provider.aws: version = "~> 2.52"

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

Let's run `tfsec .` again:

```
1 potential problems detected:

Problem 1

  [AWS004][] Resource 'aws_lb_listener.listener' uses plain HTTP instead of HTTPS.
  /workspace/aws-alb-http/listener.tf:11

       8 | 
       9 |   # Base configuration
      10 |   load_balancer_arn = aws_lb.load_balancer.arn
      11 |   protocol          = var.user_protocol    [string] "HTTP"
      12 |   port              = var.user_port
      13 | 
      14 |   # Compliant: TLS 1.2 + Forward Secrecy (strong ciphers enabled; weak ciphers disabled)

  See https://github.com/liamg/tfsec/wiki/AWS004 for more information.
```

Yay! It works! Do we even need Terraform to be initialized?

```bash
$ rm -Rf .terraform
$ tfsec .

1 potential problems detected:

Problem 1

  [AWS004][] Resource 'aws_lb_listener.listener' uses plain HTTP instead of HTTPS.
  /workspace/aws-alb-http/listener.tf:11

       8 | 
       9 |   # Base configuration
      10 |   load_balancer_arn = aws_lb.load_balancer.arn
      11 |   protocol          = var.user_protocol    [string] "HTTP"
      12 |   port              = var.user_port
      13 | 
      14 |   # Compliant: TLS 1.2 + Forward Secrecy (strong ciphers enabled; weak ciphers disabled)

  See https://github.com/liamg/tfsec/wiki/AWS004 for more information.
```

Awesome!

----

OK, let's re-clone to get a fresh copy, and then _just_ remove the ones that reference modules one at a time… (refer to original `tree -a` call at the very top).

```bash
rm -Rf examples/
tfsec .
```

It hangs again. Let's check `tests/`, and let's update the file (original/redacted version above) to remove references to _external modules_ (let's see if it can still resolve local modules)…

```hcl
# https://www.terraform.io/docs/configuration/terraform.html
terraform {
  required_version = ">= 0.12"
}

provider "aws" {
  # https://github.com/terraform-providers/terraform-provider-aws/releases/latest
  version = ">= 2.47, <3.0"
}

module "alb_http" {
  source = "../"

  # Naming
  application_name = module.aws_resource_tags.application
  environment      = module.aws_resource_tags.environment

  # General configuration
  deletion_protection = false
  enable_ipv6         = false
  external            = false

  # Tags
  tags = module.aws_resource_tags.common_tags

  # Network/VPC configuration
  vpc_id  = module.network.vpc_id
  subnets = module.network.internal_subnet_ids
  security_groups = [
    module.network.default_security_group_id,
    module.network.web_vpc_security_group_id,
  ]
}
```

Hangs again. Since this file really only includes modules and passes values into them, let's get rid of the whole file.

```bash
rm -f tests/aws_alb_test.tf
```

----

Let's run `tree -a` _one more time_ just so that we can see what's currently there.

```
.
├── .circleci
│   └── config.yml
├── .editorconfig
├── .git
│   ├── HEAD
│   ├── config
│   ├── hooks
│   ├── index
│   ├── logs
│   │   ├── HEAD
│   │   └── refs
│   │       ├── heads
│   │       │   └── master
│   │       └── remotes
│   │           └── origin
│   │               └── HEAD
│   ├── objects
│   │   ├── info
│   │   └── pack
│   │       ├── pack-4c08ece7129b7f2a5b368aaf01f5f596abf307cd.idx
│   │       └── pack-4c08ece7129b7f2a5b368aaf01f5f596abf307cd.pack
│   ├── packed-refs
│   └── refs
│       ├── heads
│       │   └── master
│       ├── remotes
│       │   └── origin
│       │       └── HEAD
│       └── tags
├── .gitignore
├── .golangci.yml
├── .markdownlintrc
├── .templates
│   ├── Makefile
│   ├── Makefile.tmpl
│   └── README.md
├── .terraform-validator.yaml
├── .tflint.hcl
├── CHANGELOG.md
├── Makefile
├── README.md
├── VERSION
├── alb.tf
├── bin
│   ├── examples.sh
│   └── show-skips.sh
├── iam.tf
├── listener.tf
├── output.tf
├── package-lock.json
├── package.json
├── target_group.tf
├── tests
│   ├── .gitignore
│   ├── aws_alb_test.go
│   ├── go.mod
│   └── go.sum
├── tf-modules-stable.png
├── variables.tf
└── versions.tf

19 directories, 41 files
```

We have a few **directories** which start with periods, a few **files** which start with periods, and **plenty** of `*.tf` files — _but none of them reference modules at all!_

Running again:

```bash
tfsec .


1 potential problems detected:

Problem 1

  [AWS004][] Resource 'aws_lb_listener.listener' uses plain HTTP instead of HTTPS.
  /Library/WebServer/Documents/terraform-modules/aws-alb-http/listener.tf:11

       8 | 
       9 |   # Base configuration
      10 |   load_balancer_arn = aws_lb.load_balancer.arn
      11 |   protocol          = var.user_protocol    [string] "HTTP"
      12 |   port              = var.user_port
      13 | 
      14 |   # Compliant: TLS 1.2 + Forward Secrecy (strong ciphers enabled; weak ciphers disabled)

  See https://github.com/liamg/tfsec/wiki/AWS004 for more information.
```

Working again.
